src/core/apparmor-setup.c

```diff
@@ -0,0 +1,99 @@
+#include <sys/apparmor.h>
+#endif
+#include <unistd.h>
+
+#include "apparmor-setup.h"
+#include "apparmor-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "macro.h"
+#include "string-util.h"
+#include "strv.h"
+
+#if HAVE_APPARMOR
+DEFINE_TRIVIAL_CLEANUP_FUNC(aa_kernel_interface *, aa_kernel_interface_unref);
+#endif
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r;
+        _cleanup_close_ int fd = -1;
+        _cleanup_(aa_kernel_interface_unrefp) aa_kernel_interface *interface = NULL;
+        const char *current_file, *profile_file;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        /* To enable LSM stacking a patch to the kernel is proposed to create a
+         * per-LSM subdirectory to distinguish between the LSMs. Therefore, we
+         * read the file from the LSM specific directory first and only if that
+         * fails the one from the generic directory.
+         */
+        FOREACH_STRING(current_file, "/proc/self/attr/apparmor/current", "/proc/self/attr/current") {
+                r = read_one_line_file(current_file, &current_profile);
+                if (r == -ENOENT)
+                        continue;
+                else if (r < 0)
+                        log_warning_errno(r, "Failed to read current AppArmor profile from file %s, ignoring: %m", current_file);
+                else
+                        break;
+        }
+        if (!current_profile) {
+                log_warning("Failed to get the current AppArmor profile of systemd from /proc/self/attr/apparmor/current or /proc/self/attr/current, ignoring.");
+                return 0;
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("We are already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        FOREACH_STRING(profile_file, "/etc/systemd/profile.apparmor", "/usr/lib/systemd/profile.apparmor") {
+                fd = open(profile_file, O_RDONLY | O_CLOEXEC);
+                if (fd < 0) {
+                        if (errno == ENOENT)
+                                continue;
+                        log_warning_errno(errno, "Failed to open binary AppArmor profile file %s, ignoring: %m", profile_file);
+                } else
+                        break;
+        }
+        /* Do not log anything if both files do not exist, because we assume the
+         * user does not want to use AppArmor for systemd itself in this case.
+         * Other errors have already been logged above.
+         */
+        if (fd < 0)
+                return 0;
+
+        r = aa_kernel_interface_new(&interface, NULL, NULL);
+        if (r < 0) {
+                log_warning_errno(errno, "Failed to get a new AppArmor kernel interface: %m");
+                return 0;
+        }
```

Comment:
This is going to fail if securityfs or apparmorfs are not premounted. Basically there is a hidden dependency here.

Reply:
mount_setup_early ensures that api filesystems like securitfy are mounted. It is called before in main.c before mac_apparmor_setup. The comment in mount_setup_early explicitly mentions this as a requirement for SELinux so it should be ok for AppArmor too.

Reply:
Yeah, we do have "basic" filesystems mounted already directly by systemd mount_setup_early, it handles things like dev, proc, sys, securityfs, for selinux, apparmor, smack, etc.

Reply:
Actually, src/core/mount-setup.c only covers securityfs + smackfs. selinuxf is mounted by selinux code. apparmorfs is currently not covered, didn't know this was a thing. Maybe this PR should carry another commit mounting it too, if that's necessary?

Reply:
Its new and hasn't been merged yet. Once it lands we can work on a patch for that.